### PR TITLE
Move modifications between groups

### PR DESCRIPTION
**Pull request by archendel**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 1]
```diff
@@ -64,24 +64,47 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
     }
 
     @Transactional // To have all move in the same transaction (atomic)
-    public void moveModifications(UUID groupUuid, List<UUID> modifications, UUID before) {
```
Comment (flomillot) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:67]: I know it was before your commits, but is it possible to rename "before"  in something more explicit ? I didn't understand what it was before reading the whole method.

**Review by flomillot — APPROVED** [review]

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java (hunk 1 above)
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:67]: fixed

**Review by sBouzols — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 2]
```diff
@@ -64,24 +64,53 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
...
+        ModificationGroupEntity originModificationGroupEntity = getGroupFromModificationUuid(modifications.get(0));
 
-        Map<UUID, ModificationEntity> originalModifications = modificationRepository.findAllBaseByGroupId(groupUuid).stream()
-            .collect(Collectors.toMap(ModificationEntity::getId, Function.identity(), (x, y) -> y, LinkedHashMap::new));
+        Map<UUID, ModificationEntity> originModifications = modificationRepository.findAllBaseByGroupId(originModificationGroupEntity.getId()).stream()
+                .collect(Collectors.toMap(ModificationEntity::getId, Function.identity(), (x, y) -> y, LinkedHashMap::new));
 
-        if (!originalModifications.keySet().containsAll(modifications) || (before != null && !originalModifications.containsKey(before))) {
-            throw new NetworkModificationException(MODIFICATION_NOT_FOUND);
+        UUID originalModificationUUID = originModificationGroupEntity.getId();
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:74]: I agree, fixed

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 3]
```diff
@@ -64,24 +64,53 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
     }
 
     @Transactional // To have all move in the same transaction (atomic)
-    public void moveModifications(UUID groupUuid, List<UUID> modifications, UUID before) {
-        /* when before == null we move at the end of list */
-        var modificationGroupEntity = getModificationGroup(groupUuid);
+    //when we move modifications, we move them right before referenceModification when it is defined, at the end of list otherwise
+    public void moveModifications(UUID destinationGroupUuid, List<UUID> modifications, UUID referenceModificationUuid) {
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:68]: > It's strange to get the groupEntity of the first modification in the list. What happen if we pass modifications from different `originGroupUUID`

There is a control checking that all modifications belong to the same group 
`            // moved modifications must all belong to the origin modification group, and before must belong to destination one
            if (!originModifications.keySet().containsAll(modifications) || referenceModificationUuid != null && !destinationModifications.containsKey(referenceModificationUuid)) {
                throw new NetworkModificationException(MODIFICATION_NOT_FOUND);
            }`

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:74]: fixed

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 4]
```diff
@@ -64,24 +64,53 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
...
+            newDestinationModificationList.addAll(index, movedModifications);
+
+            originModificationGroupEntity.setModifications(newDestinationModificationList);
+        } else {
+            ModificationGroupEntity destinationModificationGroupEntity = getModificationGroup(destinationGroupUuid);
+
+            Map<UUID, ModificationEntity> destinationModifications = modificationRepository.findAllBaseByGroupId(destinationGroupUuid).stream()
+                                                                         .collect(Collectors.toMap(ModificationEntity::getId, Function.identity(), (x, y) -> y, LinkedHashMap::new));
+
+            // moved modifications must all belong to the origin modification group, and before must belong to destination one
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:104]: fixed

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 5]
```diff
@@ -64,24 +64,53 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
...
-            throw new NetworkModificationException(MODIFICATION_NOT_FOUND);
+        UUID originalModificationUUID = originModificationGroupEntity.getId();
+        if (originalModificationUUID == null) {
+            throw new NetworkModificationException(MODIFICATION_GROUP_NOT_FOUND);
         }
 
-        List<ModificationEntity> modificationsToMove = modifications.stream().map(originalModifications::remove).collect(Collectors.toList());
+        // if moving within the same group
+        if (originalModificationUUID.equals(destinationGroupUuid)) {
+            if (!originModifications.keySet().containsAll(modifications) || referenceModificationUuid != null && !originModifications.containsKey(referenceModificationUuid)) {
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:81]: fixed

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 6]
```diff
@@ -64,24 +64,53 @@ public void saveModifications(UUID groupUuid, List<? extends ModificationEntity>
...
+                throw new NetworkModificationException(MODIFICATION_NOT_FOUND);
+            }
+
+            List<ModificationEntity> movedModifications = modifications.stream().map(originModifications::remove).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:85]: done

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 7]
```diff
@@ -33,6 +33,7 @@
 
 import static org.gridsuite.modification.server.NetworkModificationException.Type.MODIFICATION_GROUP_NOT_FOUND;
 import static org.gridsuite.modification.server.NetworkModificationException.Type.MODIFICATION_NOT_FOUND;
+import static org.gridsuite.modification.server.NetworkModificationException.Type.MOVE_MODIFICATION_ERROR;;
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:36]: remove one `;`

Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java]: fixed

**Review by sBouzols — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java [hunk 8]
```diff
@@ -96,6 +128,13 @@ public List<ModificationInfos> getModifications(List<UUID> uuids) {
             .collect(Collectors.toList());
     }
 
+    @Transactional
+    public ModificationGroupEntity getGroupFromModificationUuid(UUID modificationUuid) {
```
[the remark this answers is not in the document]
Comment (archendel) [reply src/main/java/org/gridsuite/modification/server/repositories/NetworkModificationRepository.java:132]: done

**Pull request merged**